File: 2026-04-09/trend_images/dl.js
// Download images for trend topics via CDP Google Images
const http = require('http');
const https = require('https');
const fs = require('fs');
const path = require('path');

const PROXY = 'http://localhost:3456';
const DIR = 'C:/Users/gongjue/tiktok-reports/2026-04-09/trend_images';

const TOPICS = [
  ['01', 'star_wars_maul', 'Star Wars Maul Shadow Lord animated series'],
  ['02', 'the_boys_s5', 'The Boys Season 5 TV show'],
  ['03', 'invincible_s4', 'Invincible Season 4 animated series'],
  ['04', 'daredevil_born_again', 'Daredevil Born Again Season 2'],
  ['05', 'hades_ii', 'Hades II game'],
  ['06', 'diablo_iv_hatred', 'Diablo IV Vessel of Hatred expansion'],
  ['07', 'pokemon_champions', 'Pokemon Champions game 2026'],
  ['08', 'agentic_ai', 'Agentic AI technology'],
  ['09', 'stranger_things_85', 'Stranger Things Tales from 1985'],
  ['10', 'kenny_omega', 'Kenny Omega AEW wrestling return'],
];

function get(url) {
  return new Promise((ok, fail) => {
    http.get(url, r => {
      let d = ''; r.on('data', c => d += c);
      r.on('end', () => { try { ok(JSON.parse(d)); } catch { ok(d); } });
    }).on('error', fail);
  });
}

function post(url, body) {
  return new Promise((ok, fail) => {
    const u = new URL(url);
    const req = http.request({ hostname: u.hostname, port: u.port, path: u.pathname + u.search, method: 'POST', headers: { 'Content-Type': 'text/plain' } }, r => {
      let d = ''; r.on('data', c => d += c);
      r.on('end', () => { try { ok(JSON.parse(d)); } catch { ok(d); } });
    });
    req.on('error', fail); req.write(body); req.end();
  });
}

function dl(url, fp) {
  return new Promise((ok, fail) => {
    const go = (u, n) => {
      if (n > 5) return fail(new Error('redirects'));
      const mod = u.startsWith('https') ? https : http;
      mod.get(u, { headers: { 'User-Agent': 'Mozilla/5.0' } }, r => {
        if (r.statusCode >= 300 && r.statusCode < 400 && r.headers.location) return go(r.headers.location, n + 1);
        if (r.statusCode !== 200) return fail(new Error('HTTP ' + r.statusCode));
        const s = fs.createWriteStream(fp); r.pipe(s);
        s.on('finish', () => { s.close(); ok(fs.statSync(fp).size); });
      }).on('error', fail);
    };
    go(url, 0);
  });
}

const wait = ms => new Promise(r => setTimeout(r, ms));

async function doTopic(id, slug, query) {
  console.log(`\n=== ${id}: ${query} ===`);
  const q = encodeURIComponent(query);
  const surl = `https://www.google.com/search?q=${q}&tbm=isch`;
  const tab = await get(`${PROXY}/new?url=${encodeURIComponent(surl)}`);
  const tid = tab.targetId;
  console.log('tab:', tid);
  await wait(3000);
  await get(`${PROXY}/scroll?target=${tid}&y=800`);
  await wait(1500);

  // Extract encrypted-tbn thumbnail URLs
  const js1 = `JSON.stringify([...document.querySelectorAll('img[src*="encrypted-tbn"]')].map(i=>i.src).filter((v,i,a)=>a.indexOf(v)===i).slice(0,15))`;
  const r1 = await post(`${PROXY}/eval?target=${tid}`, js1);
  let urls = [];
  try { urls = JSON.parse(r1.value); } catch {}
  console.log(`thumbnails: ${urls.length}`);

  // Also extract from scripts
  const js2 = `JSON.stringify((() => { const u=[]; document.querySelectorAll('script').forEach(s => { const m = s.textContent.match(/https?:\\/\\/[^"'\\\\s]+\\.(?:jpg|jpeg|png|webp)/gi); if(m) m.forEach(x => { if(!x.includes('google')&&!x.includes('gstatic')) u.push(x); }); }); return [...new Set(u)].slice(0,20); })())`;
  const r2 = await post(`${PROXY}/eval?target=${tid}`, js2);
  let sUrls = [];
  try { sUrls = JSON.parse(r2.value); } catch {}
  console.log(`script urls: ${sUrls.length}`);

  const all = [...urls, ...sUrls];
  let got = 0;
  const used = new Set();
  for (const u of all) {
    if (got >= 3) break;
    if (used.has(u)) continue;
    const fp = path.join(DIR, `${id}_${slug}_${got + 1}.jpg`);
    try {
      const sz = await dl(u, fp);
      if (sz >= 5000) { console.log(`  OK ${got+1}: ${sz}b`); got++; used.add(u); }
      else { fs.unlinkSync(fp); }
    } catch (e) { try { fs.unlinkSync(fp); } catch {} }
  }

  // fallback: if not enough, click first result and extract full image
  if (got < 3) {
    console.log('  trying click fallback...');
    const jsClick = `JSON.stringify((() => {
      const items = document.querySelectorAll('[data-ri]');
      const results = [];
      for (let i = 0; i < Math.min(6, items.length); i++) {
        items[i].click();
      }
      return items.length;
    })())`;
    await post(`${PROXY}/eval?target=${tid}`, jsClick);
    await wait(2000);

    // After clicking, look for full-size image in side panel
    const jsFull = `JSON.stringify([...document.querySelectorAll('img[src^="http"]')].filter(i => i.naturalWidth > 200 && !i.src.includes('google') && !i.src.includes('gstatic')).map(i => i.src).filter((v,i,a) => a.indexOf(v) === i).slice(0, 10))`;
    const r3 = await post(`${PROXY}/eval?target=${tid}`, jsFull);
    let fullUrls = [];
    try { fullUrls = JSON.parse(r3.value); } catch {}
    console.log(`  full urls: ${fullUrls.length}`);

    for (const u of fullUrls) {
      if (got >= 3) break;
      if (used.has(u)) continue;
      const fp = path.join(DIR, `${id}_${slug}_${got + 1}.jpg`);
      try {
        const sz = await dl(u, fp);
        if (sz >= 5000) { console.log(`  OK ${got+1}: ${sz}b`); got++; used.add(u); }
        else { fs.unlinkSync(fp); }
      } catch { try { fs.unlinkSync(fp); } catch {} }
    }
  }

  await get(`${PROXY}/close?target=${tid}`);
  console.log(`  result: ${got}/3`);
  return got;
}

async function main() {
  let total = 0;
  for (const [id, slug, query] of TOPICS) {
    total += await doTopic(id, slug, query);
    await wait(1000);
  }
  console.log(`\n=== TOTAL: ${total}/30 ===`);
  const files = fs.readdirSync(DIR).filter(f => /\.(jpg|png)$/i.test(f));
  for (const f of files) {
    console.log(`  ${f}: ${fs.statSync(path.join(DIR, f)).size}b`);
  }
}

main().catch(e => { console.error(e); process.exit(1); });

File: 2026-04-09/dl.py
"""Download trend topic images via CDP proxy."""
import json, os, time, urllib.request, urllib.parse

OUT = r'C:\Users\gongjue\tiktok-reports\2026-04-09\trend_images'
os.makedirs(OUT, exist_ok=True)
PROXY = 'http://localhost:3456'

def proxy_get(path):
    with urllib.request.urlopen(f'{PROXY}{path}', timeout=30) as r:
        return r.read().decode()

def proxy_post(path, body):
    req = urllib.request.Request(f'{PROXY}{path}', data=body.encode(), method='POST')
    with urllib.request.urlopen(req, timeout=30) as r:
        return r.read().decode()

def download_img(url, filepath):
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'})
        with urllib.request.urlopen(req, timeout=20) as r:
            data = r.read()
            if len(data) > 3000:
                with open(filepath, 'wb') as f:
                    f.write(data)
                return True
    except Exception:
        pass
    return False

def get_images_from_page(page_url):
    try:
        resp = json.loads(proxy_get(f'/new?url={urllib.parse.quote(page_url, safe="")}'))
        tid = resp.get('targetId') or resp.get('id')
    except Exception as e:
        print(f'    open failed: {e}')
        return []
    if not tid:
        return []
    time.sleep(3)
    try:
        proxy_get(f'/scroll?target={tid}&y=1500')
    except Exception:
        pass
    time.sleep(0.5)
    js = '''(() => {
        const urls = [];
        document.querySelectorAll('img').forEach(img => {
            const s = img.src || img.getAttribute('data-src') || img.getAttribute('data-lazy-src') || '';
            if (s.startsWith('http') && s.length > 50) {
                const w = img.naturalWidth || img.width || 0;
                const h = img.naturalHeight || img.height || 0;
                urls.push({ src: s, w, h, area: w * h });
            }
        });
        urls.sort((a, b) => b.area - a.area);
        return JSON.stringify(urls.slice(0, 10).map(u => u.src));
    })()'''
    urls = []
    try:
        result = proxy_post(f'/eval?target={tid}', js)
        urls = json.loads(result)
    except Exception:
        pass
    try:
        proxy_get(f'/close?target={tid}')
    except Exception:
        pass
    return urls

def search_google_images(query):
    search_url = f'https://www.google.com/search?q={urllib.parse.quote(query)}&tbm=isch&tbs=isz:l'
    try:
        resp = json.loads(proxy_get(f'/new?url={urllib.parse.quote(search_url, safe="")}'))
        tid = resp.get('targetId') or resp.get('id')
    except Exception as e:
        print(f'    search open failed: {e}')
        return []
    if not tid:
        return []
    time.sleep(3)
    try:
        proxy_get(f'/scroll?target={tid}&y=2000')
    except Exception:
        pass
    time.sleep(1)
    # Extract image URLs
    js = '''(() => {
        const urls = new Set();
        document.querySelectorAll('a').forEach(a => {
            const href = a.href || '';
            const m = href.match(/imgurl=([^&]+)/);
            if (m) urls.add(decodeURIComponent(m[1]));
        });
        document.querySelectorAll('img').forEach(img => {
            const s = img.src || img.getAttribute('data-src') || '';
            if (s.startsWith('http') && !s.includes('google.com') && !s.includes('gstatic.com') && !s.includes('googleapis.com') && s.length > 30) {
                urls.add(s);
            }
        });
        return JSON.stringify([...urls].slice(0, 15));
    })()'''
    urls = []
    try:
        result = proxy_post(f'/eval?target={tid}', js)
        urls = json.loads(result)
    except Exception:
        pass
    # Try clicking thumbnails for full-res
    if len(urls) < 3:
        click_js = '''(() => {
            const thumbs = document.querySelectorAll('[data-tbnid]');
            return JSON.stringify([...thumbs].slice(0, 5).map(t => t.getAttribute('data-tbnid')));
        })()'''
        tbnids = []
        try:
            tbnids = json.loads(proxy_post(f'/eval?target={tid}', click_js))
        except Exception:
            pass
        for tbnid in tbnids[:5]:
            try:
                proxy_post(f'/click?target={tid}', f'[data-tbnid="{tbnid}"]')
            except Exception:
                pass
            time.sleep(1.5)
            full_js = '''(() => {
                const imgs = [];
                document.querySelectorAll('img[src]').forEach(img => {
                    const s = img.src;
                    if (s.startsWith('http') && !s.includes('google') && !s.includes('gstatic') && !s.includes('googleapis') && (img.naturalWidth > 200 || img.width > 200)) {
                        imgs.push(s);
                    }
                });
                return JSON.stringify(imgs);
            })()'''
            try:
                big_urls = json.loads(proxy_post(f'/eval?target={tid}', full_js))
                for u in big_urls:
                    if u not in urls:
                        urls.append(u)
            except Exception:
                pass
    try:
        proxy_get(f'/close?target={tid}')
    except Exception:
        pass
    return urls

topics = [
    {'num': '01', 'slug': 'star_wars_maul',
     'queries': ['Star Wars Maul Shadow Lord animated series poster 2026'],
     'pages': ['https://www.starwars.com/news/star-wars-maul-shadow-lord-first-trailer-poster-art',
               'https://www.starwars.com/series/star-wars-maul-shadow-lord']},
    {'num': '02', 'slug': 'the_boys_s5',
     'queries': ['The Boys Season 5 final season poster 2026 Homelander'],
     'pages': ['https://heroichollywood.com/the-boys-battlelines-posters-final-season/',
               'https://cosmicbook.news/the-boys-final-season-posters-ccxp']},
    {'num': '03', 'slug': 'invincible_s4',
     'queries': ['Invincible Season 4 animated poster key art 2026'],
     'pages': ['https://bleedingcool.com/tv/invincible-season-4-key-art-poster-offers-episode-release-schedule/']},
    {'num': '04', 'slug': 'daredevil_born_again',
     'queries': ['Daredevil Born Again Season 2 poster 2026 MCU'],
     'pages': ['https://thefutureoftheforce.com/2026/03/14/the-latest-daredevil-born-again-season-2-poster-unleashes-the-kingpin/',
               'https://www.marvel.com/articles/tv-shows/daredevil-born-again-season-2-every-vigilante-character-posters']},
    {'num': '05', 'slug': 'hades_ii',
     'queries': ['Hades II game key art Melinoe Supergiant'],
     'pages': ['https://www.creativeuncut.com/art_hades-2_a.html',
               'https://www.creativeuncut.com/gallery-47/hades2-melinoe.html']},
    {'num': '06', 'slug': 'diablo_iv_hatred',
     'queries': ['Diablo IV Lord of Hatred expansion key art 2026'],
     'pages': ['https://diablo4.blizzard.com/en-us/lord-of-hatred']},
    {'num': '07', 'slug': 'pokemon_champions',
     'queries': ['Pokemon Champions game 2026 Nintendo Switch art'],
     'pages': ['https://champions.pokemon.com/en-us/']},
    {'num': '08', 'slug': 'agentic_ai',
     'queries': ['Agentic AI revolution 2026 concept illustration'],
     'pages': ['https://machinelearningmastery.com/7-agentic-ai-trends-to-watch-in-2026/']},
    {'num': '09', 'slug': 'stranger_things_85',
     'queries': ['Stranger Things Tales from 85 animated Netflix poster 2026'],
     'pages': ['https://www.justjared.com/2026/03/27/netflix-shares-new-trailer-photos-poster-for-stranger-things-tales-from-85/']},
    {'num': '10', 'slug': 'kenny_omega',
     'queries': ['Kenny Omega AEW return 2026 wrestling'],
     'pages': ['https://www.allelitewrestling.com/post/aew-dynamite-preview-april-8-2026']},
]

total = 0
for topic in topics:
    print(f"\n=== {topic['num']} {topic['slug']} ===")
    downloaded = 0
    all_urls = []

    # Try pages first
    for page_url in topic['pages']:
        if downloaded >= 3:
            break
        print(f"  Page: {page_url[:70]}...")
        urls = get_images_from_page(page_url)
        print(f"    Found {len(urls)} images")
        all_urls.extend(urls)
        for url in urls:
            if downloaded >= 3:
                break
            fn = f"{topic['num']}_{topic['slug']}_{downloaded + 1}.jpg"
            fp = os.path.join(OUT, fn)
            ok = download_img(url, fp)
            if ok:
                downloaded += 1
                sz = os.path.getsize(fp)
                print(f"    SAVED: {fn} ({sz} bytes)")

    # Google Images fallback
    if downloaded < 3:
        for query in topic['queries']:
            if downloaded >= 3:
                break
            print(f"  Google: {query}")
            urls = search_google_images(query)
            print(f"    Found {len(urls)} image URLs")
            for url in urls:
                if downloaded >= 3:
                    break
                if url in all_urls:
                    continue
                fn = f"{topic['num']}_{topic['slug']}_{downloaded + 1}.jpg"
                fp = os.path.join(OUT, fn)
                ok = download_img(url, fp)
                if ok:
                    downloaded += 1
                    sz = os.path.getsize(fp)
                    print(f"    SAVED: {fn} ({sz} bytes)")

    print(f"  Result: {downloaded}/3")
    total += downloaded
    time.sleep(0.5)

# Final report
files = sorted(os.listdir(OUT))
print(f"\n=== COMPLETE: {total}/30 images, {len(files)} files ===")
for f in files:
    sz = os.path.getsize(os.path.join(OUT, f))
    print(f"  {f} ({sz}b)")
